contracts/starknet/teleport_router.cairo

```diff
@@ -0,0 +1,297 @@
+@storage_var
+func _gateways(domain : felt) -> (res : felt):
+end
+
+# TODO
+# EnumerableSet.Bytes32Set private allDomains;
+@storage_var
+func _allDomains() -> ():
+end
+
+# address public parent;
+@storage_var
+func _parent() -> (res : felt):
+end
+
+# TokenLike immutable public dai; // L1 DAI ERC20 token
+@storage_var
+func _dai() -> (res : felt):
+end
+
+# event Rely(address indexed usr);
+@event
+func Rely(usr : felt):
+end
+
+# event Deny(address indexed usr);
+@event
+func Deny(usr : felt):
+end
+
+# event File(bytes32 indexed what, bytes32 indexed domain, address data);
+@event
+func File_ilk(what : felt, domain : felt, data : felt):
+end
+
+# event File(bytes32 indexed what, address data);
+@event
+func File(what : felt, data : felt):
+end
+
+# modifier auth {
+#     require(wards[msg.sender] == 1, "TeleportRouter/not-authorized");
+#     _;
+# }
+func auth{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+    let (caller) = get_caller_address()
+    let (ward) = _wards.read(caller)
+    with_attr error_message("l2_dai_bridge/not-authorized"):
+        assert ward = 1
+    end
+    return ()
+end
+
+# constructor(address dai_) {
+@constructor
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(dai : felt):
+    # dai = TokenLike(dai_);
+    _dai.write(dai)
+
+    let (caller) = get_caller_address()
+
+    # wards[msg.sender] = 1;
+    _wards.write(caller, 1)
+
+    # emit Rely(msg.sender);
+    Rely.emit(caller)
+
+    return ()
+end
+
+# // --- Administration ---
+# function rely(address usr) external auth {
+@external
+func rely{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(usr : felt):
+    auth()
+
+    # require(live == 1, "Vat/not-live");
+    require_live()
+
+    # wards[usr] = 1;
+    _wards.write(usr, 1)
+
+    # emit Rely(usr);
+    Rely.emit(usr)
+
+    return ()
+end
+
+# function deny(address usr) external auth {
+@external
+func deny{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(user : felt):
+    auth()
+
+    # wards[usr] = 0;
+    _wards.write(user, 0)
+
+    # emit Deny(usr);
+    Deny.emit(user)
+
+    return ()
+end
+
+# function file(bytes32 what, bytes32 domain, address data) external auth {
+@external
+func file{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    what : felt, domain : felt, data : felt
+):
+    # address prevGateway = gateways[domain];
+    let prev_gateway = _gateways.read(domain)
+
+    # if(prevGateway == address(0)) {
+    if prev_gateway == 0:
+        # if(data != address(0)) allDomains.add(domain);
+        let (is_data_zero) = is_not_zero(data)
+        if is_data_zero == 0:
+            # allDomains.add(domain); TODO
+        end
+    else:
+        # if(data == address(0)) allDomains.remove(domain);
+        if data == 0:
+            # allDomains.remove(domain); TODO
+        end
+    end
+
+    # gateways[domain] = data;
+    _gateways.write(domain, data)
```

Comment:
In next `file` function validation precedes storage write. Let's keep it consistent event if solidity version is not.